File: lib/java/opentoken/src/test/java/com/truveta/opentoken/attributes/general/DecimalAttributeTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.general;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class DecimalAttributeTest {
    private DecimalAttribute attribute;

    @BeforeEach
    void setUp() {
        attribute = new DecimalAttribute();
    }

    @Test
    void testGetName_ShouldReturnDecimal() {
        assertEquals("Decimal", attribute.getName());
    }

    @Test
    void testGetAliases_ShouldReturnDecimalAlias() {
        String[] aliases = attribute.getAliases();
        assertEquals(1, aliases.length);
        assertEquals("Decimal", aliases[0]);
    }

    @Test
    void testNormalize_ValidDecimal_ShouldReturnNormalized() {
        assertEquals("123.45", attribute.normalize("123.45"));
        assertEquals("0.0", attribute.normalize("0"));
        assertEquals("-456.78", attribute.normalize("-456.78"));
        assertEquals("789.0", attribute.normalize("  789  "));
        assertEquals("0.5", attribute.normalize(".5"));
        assertEquals("1.0", attribute.normalize("1."));
    }

    @Test
    void testNormalize_WithPositiveSign_ShouldNormalize() {
        assertEquals("123.45", attribute.normalize("+123.45"));
    }

    @Test
    void testNormalize_ScientificNotation_ShouldNormalize() {
        String result = attribute.normalize("1.5e10");
        assertEquals("1.5E10", result);

        result = attribute.normalize("-3.14E-2");
        assertEquals("-0.0314", result);
    }

    @Test
    void testNormalize_NullValue_ShouldThrowException() {
        assertThrows(IllegalArgumentException.class, () -> attribute.normalize(null));
    }

    @Test
    void testNormalize_InvalidDecimal_ShouldThrowException() {
        assertThrows(IllegalArgumentException.class, () -> attribute.normalize("abc"));
        assertThrows(IllegalArgumentException.class, () -> attribute.normalize(""));
        assertThrows(IllegalArgumentException.class, () -> attribute.normalize("   "));
    }

    @Test
    void testValidate_ValidDecimal_ShouldReturnTrue() {
        assertTrue(attribute.validate("123.45"));
        assertTrue(attribute.validate("0"));
        assertTrue(attribute.validate("-456.78"));
        assertTrue(attribute.validate("  789  "));
        assertTrue(attribute.validate("+123.45"));
        assertTrue(attribute.validate(".5"));
        assertTrue(attribute.validate("1."));
        assertTrue(attribute.validate("1.5e10"));
        assertTrue(attribute.validate("-3.14E-2"));
    }

    @Test
    void testValidate_InvalidDecimal_ShouldReturnFalse() {
        assertFalse(attribute.validate(null));
        assertFalse(attribute.validate(""));
        assertFalse(attribute.validate("abc"));
        assertFalse(attribute.validate("1.2.3"));
    }

    @Test
    void testValidate_BoundaryValues_ShouldReturnTrue() {
        assertTrue(attribute.validate("0.0"));
        assertTrue(attribute.validate("3.14159265358979"));
        assertTrue(attribute.validate("1.0E308")); // Near max double
        assertTrue(attribute.validate("1.0E-308")); // Near min positive double
    }

}

File: lib/python/opentoken/src/test/opentoken/attributes/general/decimal_attribute_test.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import pytest
from opentoken.attributes.general.decimal_attribute import DecimalAttribute


class TestDecimalAttribute:
    """Test cases for DecimalAttribute."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.attribute = DecimalAttribute()

    def test_get_name_should_return_decimal(self):
        """Test that get_name returns 'Decimal'."""
        assert self.attribute.get_name() == "Decimal"

    def test_get_aliases_should_return_decimal_alias(self):
        """Test that get_aliases returns list with 'Decimal'."""
        aliases = self.attribute.get_aliases()
        assert len(aliases) == 1
        assert "Decimal" in aliases

    def test_normalize_valid_decimal_should_normalize(self):
        """Test normalization of valid decimals."""
        assert self.attribute.normalize("123.45") == "123.45"
        assert self.attribute.normalize("0") == "0.0"
        assert self.attribute.normalize("-456.78") == "-456.78"
        assert self.attribute.normalize("  789  ") == "789.0"
        assert self.attribute.normalize("+123.45") == "123.45"

    def test_normalize_decimal_without_leading_digit_should_normalize(self):
        """Test normalization of decimals without leading digit."""
        assert self.attribute.normalize(".5") == "0.5"

    def test_normalize_decimal_without_trailing_digit_should_normalize(self):
        """Test normalization of decimals without trailing digit."""
        assert self.attribute.normalize("1.") == "1.0"

    def test_normalize_scientific_notation_should_normalize(self):
        """Test normalization of scientific notation."""
        result = self.attribute.normalize("1.5e10")
        assert result == "15000000000.0"

        result = self.attribute.normalize("-3.14E-2")
        assert result == "-0.0314"

    def test_normalize_null_value_should_raise_error(self):
        """Test that None value raises ValueError."""
        with pytest.raises(ValueError):
            self.attribute.normalize(None)

    def test_normalize_invalid_decimal_should_raise_error(self):
        """Test that invalid decimals raise ValueError."""
        with pytest.raises(ValueError):
            self.attribute.normalize("abc")
        with pytest.raises(ValueError):
            self.attribute.normalize("")
        with pytest.raises(ValueError):
            self.attribute.normalize("   ")

    def test_validate_valid_decimal_should_return_true(self):
        """Test that valid decimals validate successfully."""
        assert self.attribute.validate("123.45") is True
        assert self.attribute.validate("0") is True
        assert self.attribute.validate("-456.78") is True
        assert self.attribute.validate("  789  ") is True
        assert self.attribute.validate("+123.45") is True
        assert self.attribute.validate(".5") is True
        assert self.attribute.validate("1.") is True
        assert self.attribute.validate("1.5e10") is True
        assert self.attribute.validate("-3.14E-2") is True

    def test_validate_invalid_decimal_should_return_false(self):
        """Test that invalid decimals fail validation."""
        assert self.attribute.validate(None) is False
        assert self.attribute.validate("") is False
        assert self.attribute.validate("abc") is False
        assert self.attribute.validate("1.2.3") is False

    def test_validate_boundary_values_should_return_true(self):
        """Test that boundary values validate successfully."""
        assert self.attribute.validate("0.0") is True
        assert self.attribute.validate("3.14159265358979") is True
        assert self.attribute.validate("1.0E308") is True  # Near max double
        assert self.attribute.validate("1.0E-308") is True  # Near min positive double
